osdconfig/interface.go

```diff
@@ -0,0 +1,63 @@
+// osdconfig is a package to work with distributed config parameters
+package osdconfig
+
+import "context"
+
+// A config manager interface allows management of osdconfig parameters
+// It defines setters, getters and callback management functions
+type ConfigManager interface {
+	// GetClusterConf fetches cluster configuration data from a backend such as kvdb
+	GetClusterConf() (*ClusterConfig, error)
+
+	// Fetch node configuration data using node id
+	GetNodeConf(nodeID string) (*NodeConfig, error)
+
+	// SetClusterConf pushes cluster configuration data to the backend
+	// It is assumed that the backend will notify the implementor of this interface
+	// when a change is triggered
+	SetClusterConf(config *ClusterConfig) error
+
+	// SetNodeConf pushes node configuration data to the backend
+	// It is assumed that the backend will notify the implementor of this interface
+	// when a change is triggered
+	SetNodeConf(config *NodeConfig) error
+
+	// Register callback functions that will get triggered on a change to either
+	// cluster configuration or node configuration data in the backend.
+	// The callback function returns two channels: A channel to write to and a
+	// channel to read on. Context is provided to the callback function in order
+	// to manage cleanup during context cancellations
+	Register(name string, watcherType Watcher,
+		opt interface{},
+		cb func(ctx context.Context, opt interface{}) (chan<- *DataWrite,
+			<-chan *DataRead)) error
+
+	// Run executes callback functions, however, it is expected that users never
+	// call this function directly. Run is called behind the scenes using scheduler
+	Run(wd *DataWrite)
+
+	// GetContext returns the global osdconfig context for any related processes to use
+	GetContext() context.Context
+
+	// GetStatus returns execution status
+	// Please note that this function delivers a copy of internal status map
+	// and internal memory is cleared prior to every new callback execution cycle
+	GetStatus() map[string]*Status
+
+	// Error returns exec error if any
+	// This function should return an error during occurred previous callback execution cycle
+	Error() error
+
+	// Abort sends context cancellation to execution of callback functions
+	// This function should return immediately and not block
+	Abort()
+
+	// Wait waits while scheduler is busy processing callbacks
+	// This function should block and only return when (say) internal locks are released
+	Wait()
```

Comment:
not sure why clients need to know about this.

Reply:
Anyone wanting to set/get values in kvdb through osdconfig are recommended to use it. Also it is needed if you use Run function. I would keep it for now and remove later once use cases and stability is better understood.

Reply:
to be frank this is way complicated (because it exposes internal mechanisms), unnecessary (clients have nothing to do with this), and creates interfaces which are not required.

Reply:
per discussions yesterday, I provided a higher level util package, `configutil`. It has just 4 functions, two for creating config managers and two for setting up a watch, that's it... no context, no channels. I agree the osdconfig package is bit raw, but complexity should no longer be an issue.